File: support/web/src/main/java/com/coffee/gu/response/ApiResponse.java
package com.coffee.gu.response;


import com.coffee.gu.ErrorMessage;
import com.coffee.gu.ErrorType;

public record ApiResponse<T>(
        StatusType status,
        T data,
        ErrorMessage error
) {

    public static <T> ApiResponse<T> success() {
        return new ApiResponse<>(StatusType.SUCCESS, null, null);
    }

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(StatusType.SUCCESS, data, null);
    }

    public static <T> ApiResponse<T> error(ErrorType errorType, Object errorData) {
        return new ApiResponse<>(StatusType.ERROR, null, ErrorMessage.of(errorType, errorData));
    }

}

File: support/web/src/main/kotlin/com/coffee/gu/response/ApiResponse.kt
package com.coffee.gu.response

import com.coffee.gu.ErrorMessage
import com.coffee.gu.ErrorType

class ApiResponse<T>(
    val status: StatusType,
    val data: T? = null,
    val error: ErrorMessage? = null,
) {
    companion object {
        @JvmStatic
        fun <T> success(): ApiResponse<T> = ApiResponse(StatusType.SUCCESS)

        @JvmStatic
        fun <T> success(data: T): ApiResponse<T> = ApiResponse(StatusType.SUCCESS, data)

        @JvmStatic
        fun <T> error(errorType: ErrorType, errorData: Any? = null): ApiResponse<T> =
            ApiResponse(StatusType.ERROR, error = ErrorMessage.of(errorType, errorData))
    }
}
